src/credentials/CHIPOpCred.h

```diff
@@ -0,0 +1,184 @@
+     *
+     * @param trustedRoot  Reference to the Trusted Root ID.
+     *
+     * @return True if Trusted Root ID is in the set, false otherwise.
+     **/
+    bool IsTrustedRootIn(const CertificateKeyId & trustedRoot) const;
+
+    /**
+     * @brief Validate CHIP certificate.
+     *
+     * @param trustedRootId    Reference to the Trusted Root ID for the Certificate Set to be used
+     *                          for validation
+     * @param cert              Pointer to the CHIP certificiate to be validated. The certificate is
+     *                          required to be in this set, otherwise this function returns error.
+     * @param context           Certificate validation context.
+     *
+     * @return Returns a CHIP_ERROR on validation or other error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR ValidateCert(const CertificateKeyId & trustedRootId, const ChipCertificateData * cert, ValidationContext & context);
+
+    /**
+     * @brief Find and validate CHIP certificate.
+     *
+     * @param trustedRootId    Reference to the Trusted Root ID for the Certificate Set to be used
+     *                          for validation
+     * @param subjectDN         Subject distinguished name to use as certificate search parameter.
+     * @param subjectKeyId      Subject key identifier to use as certificate search parameter.
+     * @param context           Certificate validation context.
+     * @param cert              A pointer to the valid CHIP certificate that matches search criteria.
+     *
+     * @return Returns a CHIP_ERROR on validation or other error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR FindValidCert(const CertificateKeyId & trustedRootId, const ChipDN & subjectDN,
+                             const CertificateKeyId & subjectKeyId, ValidationContext & context, ChipCertificateData *& cert);
+
+    /**
+     * @brief A function to sign a msg using ECDSA and the respective device credentials keypair.
+     *
+     * @param msg           Message that needs to be signed
+     * @param msg_length    Length of message
+     * @param out_signature Buffer that will hold the output signature. The signature consists of: 2 EC elements (r and s),
+     *                      represented as ASN.1 DER integers, plus the ASN.1 sequence Header
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR SignMsg(const uint8_t * msg, const size_t msg_length, P256ECDSASignature & out_signature);
+
+    /**
+     * @return A pointer to device credentials (in x509 format).
+     **/
+    const uint8_t * GetDevOpCred() const { return mDeviceCredentialsDER; }
+
+    /**
+     * @return Length of the loaded device credentials buffer.
+     **/
+    uint16_t GetDevOpCredLen() const { return mDeviceCredentialsLen; }
+
+private:
+    ChipCertificateSet * mOpCreds; /**< Pointer to an array of certificate data. */
```

Comment:
What is the lifetime/ownership of these?

I see destrutor is empty, so presumably something else owns these items?
I find object ownership hard to follow when using these raw pointers.

Reply:
Yes, the initial idea is for the caller of the CASE Session to own the entire set of credentials, in order to avoid high amounts of memory copy and allocations.